### ISSUE TITLE
Docs: Confused by options for require-hyphen-before-param-description

### ISSUE THREAD

**Issue by thernstig** [gajus/eslint-plugin-jsdoc#690]
Caveat: I deleted the entire template due to this being a doc issue.

https://github.com/gajus/eslint-plugin-jsdoc/#options says:

> eslint-plugin-jsdoc options, if present, are in the form of an object supplied as the second argument in an array after the error level.

`require-hyphen-before-param-description` (https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/require-hyphen-before-param-description.md) says:

> **Options**
> This rule takes one optional string argument and an optional options object.

https://github.com/gajus/eslint-plugin-jsdoc/#require-hyphen-before-param-description also has examples that say:

> Options: ["always",{"tags":{"*":"never"}}]

This seems to contradict eachother. The first links says it should be an object. The other two indicates it is an array.

**Comment by thernstig** [comment]
For reference these are the things I have tried:

```jsonc
"jsdoc/require-hyphen-before-param-description": [
  "always",
  { "tags": { "*": "never" } }
],
// gives "Severity should be one of the following: 0 = off, 1 = warn, 2 = error (you passed '"always"')."
```

```jsonc
"jsdoc/require-hyphen-before-param-description": [
  "warn",
  ["always", { "tags": { "*": "never" } }]
],
// gives "Value ["always",{"tags":{"*":"never"}}] should be string."
```

```jsonc
"jsdoc/require-hyphen-before-param-description": [
  "warn",
  "always"
],
// gives "Value {"avoidExampleOnConstructors":true,"exemptedBy":["type"],"checkConstructors":true,"checkGetters":false,"checkSetters":false,"exemptNoArguments":false} should NOT have additional properties."
```

**Comment by thernstig** [comment]
I can also add that the table looks weird. The last "Options" seems to be half-complete. This seems to be a copy&paste from other places that also has this half-complete example. In addition, other tables with "Options" are always property names.

|||
|---|---|
|Context|everywhere|
|Tags|`param` and optionally other tags within `tags`|
|Aliases|`arg`, `argument`; potentially `prop` or other aliases|
|Recommended|false|
|Options|(a string matching `"always"|"never"`) followed by an optional object with a `tags` property|

**Comment by brettz9** [comment]
> Caveat: I deleted the entire template due to this being a doc issue.

Thanks. For the future, I've created a new template just for doc bugs.
 
> https://github.com/gajus/eslint-plugin-jsdoc/#options says:
> 
> > eslint-plugin-jsdoc options, if present, are in the form of an object supplied as the second argument in an array after the error level.
> 
> `require-hyphen-before-param-description` (https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/require-hyphen-before-param-description.md) says:
> 
> > **Options**
> > This rule takes one optional string argument and an optional options object.
> 
> https://github.com/gajus/eslint-plugin-jsdoc/#require-hyphen-before-param-description also has examples that say:
> 
> > Options: ["always",{"tags":{"*":"never"}}]
> 
> This seems to contradict eachother. The first links says it should be an object. The other two indicates it is an array.

Yes, thank you. I've amended the docs to indicate that this is only generally the case. Some rules, following the `eslint` practice, do indeed have a string (currently, just "always" or "never") as the first option.

> For reference these are the things I have tried:
> <snip>
> 
> ```js
> "jsdoc/require-hyphen-before-param-description": [
>   "warn",
>   "always"
> ],
> // gives "Value {"avoidExampleOnConstructors":true,"exemptedBy":["type"],"checkConstructors":true,"checkGetters":false,"checkSetters":false,"exemptNoArguments":false} should NOT have additional properties."
> ```

While your other expected warnings are expected (which hopefully the docs now clarify), this last result does not seem it should be happening. You might want to check this latter example again, because I would expect this would work. If not, you can file a regular bug report about it.

> I can also add that the table looks weird. The last "Options" seems to be half-complete. This seems to be a copy&paste from other places that also has this half-complete example. In addition, other tables with "Options" are always property names.
> 	
> Context 	everywhere
> Tags 	`param` and optionally other tags within `tags`
> Aliases 	`arg`, `argument`; potentially `prop` or other aliases
> Recommended 	false
> Options 	(a string matching `"always"

Thank you. This is now fixed. Was a Markdown escaping issue. As far as property names, we don't want property names here because this is a case where the first option is indeed a string, so there is no property to list.
 
> I find it confusing that all examples on https://github.com/gajus/eslint-plugin-jsdoc/#require-hyphen-before-param-description add `// Options: [{"exemptTagContexts":[{"tag":"typedef","types":true}]}]` as an array when the options is always an object in ESLint.

While following the approach of ESLint testing tools is not necessarily user friendly, and we therefore might consider stripping off the `[` and `]`, leading to something like `Options: "always",{"tags":{"*":"never"}}`, I wonder whether the array syntax better clarifies and reminds people to use an array when specifying options (since needed for adding the likes of `['warn', 'always']`). Maybe we could change it to the following to more clearly suggest the structure:

`Options: ["error"|"warn", "always", {"tags":{"*":"never"}}]`

I'm marking the issue as fixed since the problematic items are now handled, but feel free to comment further on this latter question.

**Comment by thernstig** [comment]
First of, thank you for answering so elegantly. Very much appreciate, it all makes sense now.

https://eslint.org/docs/user-guide/configuring/rules#configuring-rules says:

> If a rule has additional options, you can specify them using array literal syntax, such as:

So it calls the entire array "Options" in this case.

What threw me of in the documentation here https://github.com/gajus/eslint-plugin-jsdoc#options was:

> `eslint-plugin-jsdoc` options, if present, are generally in the form of an object 

and

> `// The options vary by rule, but are added to an options object:`

ESLint makes it sound like even "off/error/warn" are part of the options array. So I think your last example might help here. But also as you say, re-write the options chapter slightly.

You are right, this works. I made a mistake before:

````jsonc
"jsdoc/require-hyphen-before-param-description": [
  "warn",
  "always"
],
```

**Comment by brettz9** [comment]
> it all makes sense now.

Glad to hear it.
  
> What threw me of in the documentation here https://github.com/gajus/eslint-plugin-jsdoc#options was:
> 
> <snip>
> 
> > `// The options vary by rule, but are added to an options object:`

Ah, good catch there too. I've amended it to say:

> // The options vary by rule, but are generally added to an options
> //  object as follows:

**Comment by brettz9** [comment]
> But also as you say, re-write the options chapter slightly.

I guess you mean eslint's docs on configuration. We are independent from them, so you'd have to report a desire for changes at https://github.com/eslint/eslint/issues (or communicate with their developers by their [Discord chat](https://eslint.org/chat)).

**Comment by brettz9** [comment]
> ESLint makes it sound like even "off/error/warn" are part of the options array. So I think your last example might help here. 

@gajus, are you ok with my changing the README output from producing the likes of:

> Options: ["always", {"tags":{"*":"never"}}]

to:

> Options: ["error"|"warn", "always", {"tags":{"*":"never"}}]

?

**Comment by thernstig** [comment]
> I guess you mean eslint's docs on configuration. We are independent from them, so you'd have to report a desire for changes at https://github.com/eslint/eslint/issues (or communicate with their developers by their Discord chat).

I meant https://github.com/gajus/eslint-plugin-jsdoc#options

Shouldn't this line also get an update:

>eslint-plugin-jsdoc options, if present, are generally in the form of an object supplied as the second argument in an array after the error level (any exceptions to this format are explained within that rule's docs).

That one makes it sound like the options is _only_ the object, when in fact it is more. Even including `"error"|warn"`. As https://eslint.org/docs/user-guide/configuring/rules#configuring-rules says:

> If a rule has **_additional_** options

The term _additional_ makes it sound like everything for a rule is considered "options".

**Comment by brettz9** [comment]
> > I guess you mean eslint's docs on configuration. We are independent from them, so you'd have to report a desire for changes at https://github.com/eslint/eslint/issues (or communicate with their developers by their Discord chat).
> 
> I meant https://github.com/gajus/eslint-plugin-jsdoc#options
> 
> Shouldn't this line also get an update:
> 
> > eslint-plugin-jsdoc options, if present, are generally in the form of an object supplied as the second argument in an array after the error level (any exceptions to this format are explained within that rule's docs).
> 
> That one makes it sound like the options is _only_ the object, when in fact it is more. Even including `"error"|warn"`. 

Yes, indeed, I fixed that, but you might not have seen it by the time I sent the email as I had forgotten to push the changes to the repository. But I have clarified it to say:

> // The options vary by rule, but are generally added to an options
> //  object as follows:

> As https://eslint.org/docs/user-guide/configuring/rules#configuring-rules says:
> 
> > If a rule has **_additional_** options
> 
> The term _additional_ makes it sound like everything for a rule is considered "options", including `"error"|"warn"`.

This is speaking about inline config, but config can also be within JavaScript or JSON, and in such cases, it must always be an array (unless merely indicating the error/warn/off status). I think they are using "options" here more loosely to mention that the rule's configuration can *option*ally be reported as an error or a mere warning, as well as being specifically configured (e.g., to insist on double quotes with "double"). But in other places, e.g., at https://eslint.org/docs/rules/no-unused-vars#options , where in the "Options" section at "This rule takes one argument which can be a string or an object.", they are speaking of the one "argument" here as a string/object and in other rules too, treat "options" more as the individual configuration.

**Comment by thernstig** [comment]
I see, and I agree they seem to use it more loosely there. What about what your previous comment then?

> While following the approach of ESLint testing tools is not necessarily user friendly, and we therefore might consider stripping off the [ and ], leading to something like Options: "always",{"tags":{"*":"never"}}, I wonder whether the array syntax better clarifies and reminds people to use an array when specifying options (since needed for adding the likes of ['warn', 'always']). Maybe we could change it to the following to more clearly suggest the structure:

I think stripping of `[` and `]` then makes sense. As the options is never an array. Your example `['warn', 'always']` is per your definition then not not the entire options, only the `always` part is.

This is what confused me from the beginning with all examples, as options are usually  just an object (but can also have an optional string as you so clearly also pointed out by now.)

**Comment by brettz9** [comment]
> I see, and I agree they seem to use it more loosely there. What about what your previous comment then?
> 
> > While following the approach of ESLint testing tools is not necessarily user friendly, and we therefore might consider stripping off the [ and ], leading to something like Options: "always",{"tags":{"*":"never"}}, I wonder whether the array syntax better clarifies and reminds people to use an array when specifying options (since needed for adding the likes of ['warn', 'always']). Maybe we could change it to the following to more clearly suggest the structure:
> 
> I think stripping of `[` and `]` then makes sense. As the options is never an array. Your example `['warn', 'always']` is per your definition then not not the entire options, only the `'always'` part is.

Yes, I agree and think that is a good idea. I've suggested it to @gajus who is the owner of this project and set up the docs originally, so seeking his approval.

> **EDIT:**
> But then doing this is also confusing:
> `Options: ["error"|"warn", "always", {"tags":{"*":"never"}}]`
> 
> As the `"error"|"warn"` is _not_ an option.

We could use more plain English

> Options: ["error" OR "warn", ...

But the pipe symbol `|` is pretty often used in technical docs to delineate a series of alternatives, so in some ways I think that is still superior as proposed.

**Comment by thernstig** [comment]
I meant more that `"error"|"warn"` is not part of the options (not talking about the pipe, which I think is fine as that is how e.g. TypeScript and JSDoc does it.)

E.g. from here:
https://eslint.org/docs/rules/no-unused-vars#options

> This rule takes one argument which can be a string or an object.

Then saying:

> `Options: ["error"|"warn", "always", {"tags":{"*":"never"}}]`

It seems incorrect to include `"error"|"warn"` there.

Or maybe I misunderstood your answer? :)

**Comment by brettz9** [comment]
> I meant more that `"error"|"warn"` is not part of the options (not talking about the pipe, which I think is fine as that is how e.g. TypeScript and JSDoc does it.)
> 
> E.g. from here:
> https://eslint.org/docs/rules/no-unused-vars#options
> 
> > This rule takes one argument which can be a string or an object.
> 
> Then saying:
> 
> > `Options: ["error"|"warn", "always", {"tags":{"*":"never"}}]`
> 
> It seems incorrect to include `"error"|"warn"` there.
> 
> Or maybe I misunderstood your answer? :)

There are two ways to understand "options" and either is, in a sense correct. One way is with `"error"|"warn"` and another is without it.

However, maybe even less ambiguous would be:

`jsdoc/the-rule-name`: ["error"|"warn", "always", {"tags":{"*":"never"}}]`

That would also allow easier copy-pasting to adapt.

**Comment by thernstig** [comment]
That example would definitely be the most understandable, and copy-pasteable. I like it.

**Comment by thernstig** [comment]
@brettz9 Since this was closed, should there be a separate issue about https://github.com/gajus/eslint-plugin-jsdoc/issues/690#issuecomment-778159897? (would somewhat be a shame since this thread so specifically explains our reasoning back&forth)

**Comment by brettz9** [comment]
@thernstig : There is generally no need for worry on tracking small, easy issues like this. I tend to keep track of such things myself. If it is an older issue, also, it has probably stalled because of a lack of personal priority and/or energy to get to it. Just wanted approval in this case to change, as it is more of a project presentation issue, and we're good. I've made the change in https://github.com/gajus/eslint-plugin-jsdoc/commit/8f5215cfe964e374ad270adc42c11ab321c5b46f .
